File: rust/src/day04.rs
use lazy_static::lazy_static;
use regex::Regex;
use regex::RegexSet;

#[aoc_generator(day4)]
pub fn parse_input(input: &str) -> Vec<String> {
    input.split("\n\n").map(|l| l.replace("\n", " ")).collect()
}

lazy_static! {
    static ref BYR_REGEX: Regex = Regex::new(r"byr:(\d{4})\b").unwrap();
    static ref IYR_REGEX: Regex = Regex::new(r"iyr:(\d{4})\b").unwrap();
    static ref EYR_REGEX: Regex = Regex::new(r"eyr:(\d{4})\b").unwrap();
    static ref HGT_REGEX: Regex = Regex::new(r"hgt:(\d{2,3})(cm|in)\b").unwrap();
    static ref HCL_REGEX: Regex = Regex::new(r"hcl:#[0-9a-f]{6}\b").unwrap();
    static ref ECL_REGEX: Regex = Regex::new(r"ecl:(amb|blu|brn|gry|grn|hzl|oth)\b").unwrap();
    static ref PID_REGEX: Regex = Regex::new(r"pid:\d{9}\b").unwrap();
    static ref PASSPORT_CHECK: RegexSet =
        RegexSet::new(&[r"byr:", r"iyr:", r"eyr:", r"hgt:", r"hcl:", r"ecl:", r"pid:",]).unwrap();
}

fn check_passport(passport: &str) -> bool {
    PASSPORT_CHECK.matches(passport).into_iter().count() == PASSPORT_CHECK.len()
}

fn full_check(passport: &str) -> bool {
    if let Some(caps) = BYR_REGEX.captures(passport) {
        let year = caps.get(1).unwrap().as_str().parse::<usize>().unwrap();
        if !(year >= 1920 && year <= 2002) {
            return false;
        }
    } else {
        return false;
    };
    if let Some(caps) = IYR_REGEX.captures(passport) {
        let year = caps.get(1).unwrap().as_str().parse::<usize>().unwrap();
        if !(year >= 2010 && year <= 2020) {
            return false;
        }
    } else {
        return false;
    };
    if let Some(caps) = EYR_REGEX.captures(passport) {
        let year = caps.get(1).unwrap().as_str().parse::<usize>().unwrap();
        if !(year >= 2020 && year <= 2030) {
            return false;
        }
    } else {
        return false;
    };
    if let Some(caps) = HGT_REGEX.captures(passport) {
        let height = caps.get(1).unwrap().as_str().parse::<usize>().unwrap();
        let test = match caps.get(2).unwrap().as_str() {
            "cm" => height >= 150 && height <= 193,
            "in" => height >= 59 && height <= 76,
            _ => return false,
        };
        if !test {
            return false;
        }
    } else {
        return false;
    };

    HCL_REGEX.is_match(passport) && ECL_REGEX.is_match(passport) && PID_REGEX.is_match(passport)
}

#[aoc(day4, part1)]
pub fn solve_part1(input: &Vec<String>) -> usize {
    input
        .iter()
        .filter(|&passport| check_passport(passport))
        .count()
}

#[aoc(day4, part2)]
pub fn solve_part2(input: &Vec<String>) -> usize {
    input
        .iter()
        .filter(|&passport| full_check(passport))
        .count()
}

File: clj/advent2020/src/jjcomer/day04.clj
(ns jjcomer.day04
  (:require [clojure.string :as str]
            [jjcomer.util :as util]))

(def simple-check
  [#"byr:" #"iyr:" #"eyr:" #"hgt:" #"hcl:" #"ecl:" #"pid:"])

(defn do-simple-check [passport]
  (every? #(re-find % passport) simple-check))

(defn parse-input [input]
  (map #(str/replace % "\n" " ") (str/split input #"\n\n")))

(defn solve-part-1 []
  (let [passports (parse-input (util/get-input 4 :lines false))]
    (count (filter do-simple-check passports))))

(defn complex-test [passport]
  (and (re-find #"hcl:#[0-9a-f]{6}\b" passport)
       (re-find #"ecl:(amb|blu|brn|gry|grn|hzl|oth)\b" passport)
       (re-find #"pid:\d{9}\b" passport)
       (if-let [[_ year] (re-find #"byr:(\d{4})\b" passport)]
         (<= 1920 (Integer. year) 2002))
       (if-let [[_ year] (re-find #"iyr:(\d{4})\b" passport)]
         (<= 2010 (Integer. year) 2020))
       (if-let [[_ year] (re-find #"eyr:(\d{4})\b" passport)]
         (<= 2020 (Integer. year) 2030))
       (if-let [[_ height unit] (re-find #"hgt:(\d{2,3})(cm|in)\b" passport)]
         (case unit
           "cm" (<= 150 (Integer. height) 193)
           "in" (<= 59 (Integer. height) 76)
           :else false))))

(defn solve-part-2 []
  (let [passports (parse-input (util/get-input 4 :lines false))]
    (count (filter complex-test passports))))
